### ISSUE TITLE
[nombis] Releasing unused candidates vs. keeping them forever

### ISSUE THREAD

**Issue by emcho** [juberti/draughts#13]
Currently the draft seems to be making the soft assumption that no candidates are ever released. This is a pretty important point whether or not this assumption is made would have important implications. 

There are basically three options:

A) We NEVER release any candidates even if they are never used.
B) Releasing candidates is possible but has to be explicitly signalled through _signalling_ path.
C) Everything, except for the candidates used in the valid pairs gets implicitly released 

Option A might make some people unhappy as it means increased port usage on CGNs and TURN servers and it means that it is impossible for applications to turn off some interfaces for energy optimization purposes. 

It also  means that cases where some interfaces disconnect for whatever reasons (e.g. wi-fi signal dies, cable pulled out, etc.).

Option B would mean that we need new signalling for deleting candidates. It may also be a somewhat inefficient option because a key point of nombis is to improve performance and introducing more signalling is going to achieve the exact opposite.

Option C would imply the complexity of having two modes: one where checks, trickling and discovery are going on and another where they aren't. It would also mean having to use ICE resumes (intentionally avoiding use of the word restart :) )

**Comment by juberti** [comment]
I don't think these are the only options. We can do B), except with releasing candidates via the media path.

**Comment by juberti** [comment]
IETF 92 result - continue to do continuing consent on all paths, add some DONT-USE-CANDIDATE attr for controlling side to tell controlled side it is not going to continue to use a path.

**Comment by emcho** [comment]
What I took out was the above rough consensus about the very specific semantics of how to valid usable pairs you don't want to use any more. 

I don't think there were decisions (or even discussion) on all the other points in this issue.

**Comment by juberti** [comment]
I was just noting it for when I next update the draft. My current thinking:
1) Initial nomination works like regular nomination
2) Then, on controlling side: non-valid pairs are pruned after N seconds of being down (perhaps same as consent checks)
3) Controlling or controlled side may also prune any candidates that they do not wish to keep due to cost, power, or other reasons; controlled side may simply drop them; controlling side may send DONT-USE-CANDIDATE on the relevant pairs
4) If all pairs on a candidate are pruned, the candidate is pruned, unless it is the base for other activate candidates
5) New candidates may be discovered from various interface events: interface changes; captive portal events - and should cause new candidate gathering on the affected interfaces
6) New candidates may also be discovered from learning about new TURN servers, and should cause new TURN candidates to be gathered from all active interfaces
7) These new candidates are trickled ASAP, and are matched up with existing candidates to form new pairs, and processing continues at 2

**Comment by emcho** [comment]
@JonathanLennox  sorry if the discussion was  boring you ;)

@juberti most of this sounds fine, with the following comments and caveats

> 1) Initial nomination works like regular nomination

+1

> 2) Then, on controlling side: non-valid pairs are pruned after N
> seconds of being down (perhaps same as consent checks)

OK. This is *very* important though as I think the draft currently relies on those candidates being kept around. I'll open separate tickets for this as soon as I get off the plane.

What it means is that you will need to have two modes of operation: "checking" and "cruise". This is important because when you return to "checking" from "cruise" you will need to regather all the candidates you originally pruned, as they may now work. Or in other words, you will need to have something similar to an ICE restart (or resume if you prefer).

Also I don't get why you need consent checks to prune. You are doing connectivity checks so you can prune as soon as your STUN transactions fail.

> 3) Controlling or controlled side may also prune any candidates that
> they do not wish to keep due to cost, power, or other reasons;
> controlled side may simply drop them; controlling side may send
> DONT-USE-CANDIDATE on the relevant pairs

if "simply drop" means closing the socket, then this may lead to the controlling side spending an entire STUN transaction to notify the controlled as it will be waiting for an ACK that will never come. 

I suggest allowing both of them to send DONT-USE-CANDIDATE and make them wait for their server transactions to expire.

> 4) If all pairs on a candidate are pruned, the candidate is pruned,
> unless it is the base for other activate candidates

OK, so that's sort of like 5245.

> 5) New candidates may be discovered from various interface events:
> interface changes; captive portal events - and should cause new
> candidate gathering on the affected interfaces

We need to be careful here. A VPN interface going up may actually take down some of your existing interfaces because it messes with the routing tables. A Wi-Fi interface going down may take down the VPN interface that was using it. 

Bottom line is that it is dangerous to make assumptions about exactly how an event on one interface would impact your connectivity altogether, hence my comment on point 7.

> 6) New candidates may also be discovered from learning about new TURN
> servers, and should cause new TURN candidates to be gathered from all
> active interfaces

OK.

> 7) These new candidates are trickled ASAP, and are matched up with
> existing candidates to form new pairs, and processing continues at 2

So what I think would be best here, would be to have agents re-run all checks, which is another reason to have an ICE restart/resume.
